File: Frontend/BandFounder/src/types/Listing.ts
import {MusicianSlot} from "./MusicianSlot";
import {Account} from "./Account";

export interface Listing {
    id: string;
    ownerId: string;
    name: string;
    genre: string;
    type: ListingType;
    description: string;
    musicianSlots: MusicianSlot[];
    owner: Account;
}

export interface ListingWithScore {
    listing: Listing;
    similarityScore: number;
}

export interface ListingsFeedDto {
    listings: ListingWithScore[];
}

export interface ListingFeedFilters {
    excludeOwn: boolean | undefined,
    matchMusicRole: boolean | undefined,
    fromLatest: boolean | undefined,
    listingType: ListingType | undefined,
    genre: string | undefined
    pageNumber: number | undefined;
    pageSize: number | undefined;
}

export enum ListingType {
    Band = 'Band',
    CollaborativeSong = 'CollaborativeSong'
}

export function castToListingType(value: string): ListingType | null {
    return isListingType(value) ? (value as ListingType) : null;
}

function isListingType(value: string): value is ListingType {
    return Object.values(ListingType).includes(value as ListingType);
}

File: Backend/BandFounder/BandFounder.Domain/Entities/Listing.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using System.Runtime.Serialization;

namespace BandFounder.Domain.Entities;

public class Listing : Entity
{
    [Key]
    public Guid Id { get; set; }

    public required string Name { get; set; }
    
    public required Guid OwnerId { get; set; }
    public virtual Account Owner { get; set; } = null!;

    public string? GenreName { get; set; }
    public virtual Genre? Genre { get; set; }

    [Column(TypeName = "text")]
    public required ListingType Type { get; set; }

    public virtual List<MusicianSlot> MusicianSlots { get; set; } = [];

    public string? Description { get; set; }
    
    public DateTime DateCreated { get; set; } = DateTime.UtcNow;
}

public enum ListingType
{
    [EnumMember(Value = "Active")] Band,
    [EnumMember(Value = "Active")] CollaborativeSong
}